transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -161,8 +168,118 @@ def _trigger_rpc(self, device_info):
+        if not self._jlink.halted():
+            self._jlink.halt()
+        self._jlink.register_write(register, value)
+
+    def _get_register_handle(self, register_string):
+        reg_list = self._jlink.register_list()
+        reg_str_paren = '(' + register_string + ')'
+
+        for reg in reg_list:
+            reg_name = self._jlink.register_name(reg)
+            if reg_name is register_string or reg_str_paren in reg_name:
+                return reg
+        return None
+
+    def _get_ffd_max_read_length(self):
+        max_read_length = self._jlink.memory_read16(ffd_cfg.ffd_ram_addrs['max_read_length'], 1)
+        return max_read_length[0]
+
+    def _write_ffd_dump_cmd(self, start, length):
+        # start must be a 32 bit address
+        start_bytes = start.to_bytes(4, byteorder='little')
+        # length must be 16 bit length
+        length_bytes = length.to_bytes(2, byteorder='little')
+
+        self._jlink.memory_write8(ffd_cfg.ffd_ram_addrs['read_cmd.address'], start_bytes)
+        written_start = self._jlink.memory_read32(ffd_cfg.ffd_ram_addrs['read_cmd.address'], 1)
+
+        # check if start address written to RAM was correct
+        if written_start[0].to_bytes(4, byteorder='little') != start_bytes:
+            raise ArgumentError("FFD dump command address was not successfully written.")
+        else:
+            logger.info("FFD dump command address written to %s", hex(start))
+
+        self._jlink.memory_write8(ffd_cfg.ffd_ram_addrs['read_cmd.length'], length_bytes)
+        written_length = self._jlink.memory_read16(ffd_cfg.ffd_ram_addrs['read_cmd.length'], 1)
+
+        # check if data length written to RAM was correct
+        if written_length[0].to_bytes(2, byteorder='little') != length_bytes:
+            raise ArgumentError("FFD dump command length was not successfully written.")
+        else:
+            logger.info("FFD dump command length written to %d", length)
+
+    def _read_ffd_dump_resp(self, length):
+        buffer_addr = self._jlink.memory_read32(ffd_cfg.ffd_ram_addrs['read_resp.buffer_addr'], 1)
+        logger.info("Response buffer at %s", hex(buffer_addr[0]))
+        flash_dump = self._read_memory(buffer_addr[0], length)
+        return flash_dump
+
+    def _debug_rd_mem(self, _device_info, _control_info, args, _progress_callback):
+        memory = self._read_memory(args.get('address'), args.get('length'))
+        return memory
+
+    def _debug_wr_mem(self, _device_info, _control_info, args, _progress_callback):
+        nbytes = self._jlink.memory_write(args.get('address'), args.get('data'))
+        return int(nbytes)
+
+    def _dump_memory(self, device_info, _control_info, args, _progress_callback):
+        memory_type = args.get('memory')
+        start_addr  = args.get('start')
+        data_length = args.get('length')
+
+        if memory_type.lower() == 'ram':
+            memory = self._read_memory(device_info.ram_start, device_info.ram_size)
```

Comment:
This doesn't let the user dump less than the complete ram even if they specify an explicit length.  I would refactor to use `data_length` is not None and use the full region size if it is None.

Reply:
Fixed.